src/modm/platform/can/stm32-fdcan/can.cpp.in

```diff
@@ -192,13 +195,67 @@ modm::platform::Fdcan{{ id }}::initializeWithPrescaler(
+	{{ reg }}->XIDFC |= MessageRam::Config.extendedFilterCount << FDCAN_XIDFC_LSE_Pos;
+	{{ reg }}->RXF0C |= MessageRam::Config.rxFifo0Elements << FDCAN_RXF0C_F0S_Pos;
+	{{ reg }}->RXF1C |= MessageRam::Config.rxFifo1Elements << FDCAN_RXF1C_F1S_Pos;
+	// NB: RX buffer has no size configuration register; a write index is put
+	// into filter settings if desired. We don't support RX buffers.
+	{{ reg }}->TXEFC |= MessageRam::Config.txEventFifoElements << FDCAN_TXEFC_EFS_Pos;
+	{{ reg }}->TXBC |= (MessageRam::Config.txFifoElements << FDCAN_TXBC_TFQS_Pos) | (0 << FDCAN_TXBC_NDTB_Pos);
+	// Trigger memory elements unused, left at 0 (only supported on FDCAN1)
+	// {{ reg }}->TTTMC |= 0 << FDCAN_TTTMC_TME_Pos;
+
+	// Configure number of message data bytes stored in each message entry (maximum length)
+	{{ reg }}->TXESC |= static_cast<uint32_t>(MessageRam::Config.txBufferDataBytesMax) << FDCAN_TXESC_TBDS_Pos;
+	{{ reg }}->RXESC |=
+		(static_cast<uint32_t>(MessageRam::Config.rxBufferDataBytesMax) << FDCAN_RXESC_RBDS_Pos)
+		| (static_cast<uint32_t>(MessageRam::Config.rxFifo0DataBytesMax) << FDCAN_RXESC_F0DS_Pos)
+		| (static_cast<uint32_t>(MessageRam::Config.rxFifo1DataBytesMax) << FDCAN_RXESC_F1DS_Pos);
+
+	// Configure base address of each RAM section
+	{{ reg }}->SIDFC |= (MessageRam::Offset + MessageRam::Config.standardFilterListSectionOffset()) << FDCAN_SIDFC_FLSSA_Pos;
+	{{ reg }}->XIDFC |= (MessageRam::Offset + MessageRam::Config.extendedFilterListSectionOffset()) << FDCAN_XIDFC_FLESA_Pos;
+	{{ reg }}->RXF0C |= (MessageRam::Offset + MessageRam::Config.rxFifo0SectionOffset()) << FDCAN_RXF0C_F0SA_Pos;
+	{{ reg }}->RXF1C |= (MessageRam::Offset + MessageRam::Config.rxFifo1SectionOffset()) << FDCAN_RXF1C_F1SA_Pos;
+	// {{ reg }}->RXBC |= 0 << FDCAN_RXBC_RBSA_Pos;
+	{{ reg }}->TXEFC |= (MessageRam::Offset + MessageRam::Config.txEventSectionOffset()) << FDCAN_TXEFC_EFSA_Pos;
+	{{ reg }}->TXBC |= (MessageRam::Offset + MessageRam::Config.txFifoSectionOffset()) << FDCAN_TXBC_TBSA_Pos;
+	// {{ reg }}->TTTMC |= 0 << FDCAN_TTTMC_TMSA_Pos;
+
+	static_assert(MessageRam::Config.txFifoSectionOffset() + MessageRam::Config.txFifoSectionWords() == MessageRam::Config.totalSectionWords());
+	static_assert(MessageRam::Offset + MessageRam::Config.totalSectionWords() <= {{message_ram_length_words}});
+
+%% else
+	%# Series other than STM32H7 have a fixed message RAM layout.
+
+	// NB: Before STM32H7, LSS and LSE determined the number of enabled filters;
+	// other filters were present but ignored. In STM32H7, LSS and LSE determine
+	// how many filters are present in the RAM. For pre-STM32H7 series we use
+	// the max value and zero out (disable) all filters on initialization.
+	{{ reg }}->RXGFC |= (MessageRam::Config.standardFilterCount << FDCAN_RXGFC_LSS_Pos) | (MessageRam::Config.extendedFilterCount << FDCAN_RXGFC_LSE_Pos);
+%%endif
 
 	// Tx buffer: queue mode
-	{{ reg }}->TXBC = FDCAN_TXBC_TFQM;
+	{{ reg }}->TXBC |= FDCAN_TXBC_TFQM;
```

Comment:
Can anyone comment on why we enable queue mode? It leads to a weird behavior where the in-memory/SW-managed queue is a true FIFO but the HW is arbitrating according to message ID priority. In fact, this breaks the fdcan buffer test, which expects to receive messages in the order they were enqueued.

I see some related discussion here: https://github.com/modm-io/modm/discussions/710.

My opinion is that this behavior is undesireable so long as the SW-managed queue remains. I would _prefer_ to disable it.

Reply:
The behaviour is rather weird with the software queue and disabling it seems reasonable. 

Furthermore, we should allow to completely bypass the software queues. In that case I'd prefer to have the arbitration. With the 10 kB message RAM on H7 using the hardware queues only is sufficient for many applications. The code is already there in the driver, we only need to allow users to configure the software buffers as zero length: https://github.com/chris-durand/modm/commit/ddf3d179e0a8c89047f86770c009fb4f04a52ecf

Reply:
Should we somehow add an option to enable queue mode? In case the additional software queue isn't used it seems to be a useful feature.